### PR TITLE
Added Unit Tests

### PR DESCRIPTION
**Pull request by ameknas-phac**
Added unit tests for utils, mcluster, and multi level clustering

### CODE REVIEW

File: .gitignore
Comment (mattheww95) [inline .gitignore:1]: Looks good, we can hold off merging the PR till the other tests are completed.

**Review by mattheww95 — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (ameknas-phac) [reply .gitignore:1]: ok

**Comment by kylacochrane** [comment]
**Out of 21 tests, 18 are passing with pytest.**

However, three tests are still failing, and they all reside in the test_GAS_matrix_splitter.py file. All three failures occur within the same function, prep_batch_ranges, in the matrix_splitter.py class:

```
def prep_batch_ranges(self):
        self.num_batches = int(self.num_lines / self.batch_size)
        rem = self.batch_size % self.num_batches
        ranges = []
        for i in range(0,self.num_batches):
            ranges.append(i*self.batch_size,i*self.batch_size+self.batch_size)
        if rem != 0:
            r = ranges[-1]
            r[1] = self.num_lines
        self.ranges = ranges
```

**Error:**
```
>           ranges.append(i*self.batch_size,i*self.batch_size+self.batch_size)
E           TypeError: list.append() takes exactly one argument (2 given)
genomic_address_service/classes/matrix_splitter.py:54: TypeError
```

**Suggested Fix:**
To resolve this issue, the `append()` method should receive a single argument, which in this case is a tuple containing both values. The corrected line in `matrix_splitter.py` should be:

`ranges.append((i*self.batch_size, i*self.batch_size+self.batch_size))`

**Comment by sgsutcliffe** [comment]
> **Out of 21 tests, 18 are passing with pytest.**
> 
> However, three tests are still failing, and they all reside in the `test_GAS_matrix_splitter.py` file. All three failures occur within the same function, `prep_batch_ranges`, in the `matrix_splitter.py` class:
> 
> ```
> def prep_batch_ranges(self):
>         self.num_batches = int(self.num_lines / self.batch_size)
>         rem = self.batch_size % self.num_batches
>         ranges = []
>         for i in range(0,self.num_batches):
>             ranges.append(i*self.batch_size,i*self.batch_size+self.batch_size)
>         if rem != 0:
>             r = ranges[-1]
>             r[1] = self.num_lines
>         self.ranges = ranges
> ```
> 
> **Error:**
> 
> ```
> >           ranges.append(i*self.batch_size,i*self.batch_size+self.batch_size)
> E           TypeError: list.append() takes exactly one argument (2 given)
> genomic_address_service/classes/matrix_splitter.py:54: TypeError
> ```
> 
> **Suggested Fix:** A possible resolution for this issue, could be to have the `append()` method receive a single argument, which in this case is a tuple containing both values. The corrected line in `matrix_splitter.py` could be:
> 
> `ranges.append((i*self.batch_size, i*self.batch_size+self.batch_size))`

That solved that issue! @kylacochrane 41bdffe

**Comment by sgsutcliffe** [comment]
The `matrix_splitter` class was not actually being used by GAS so I removed it and the related tests. All current pytests should be passing without any issue!

**Review by apetkau — CHANGES_REQUESTED** [review]
@sgsutcliffe thanks so much for all the great work you've put into getting these tests to work 😄. I really appreciate it. Along with your work @kylacochrane 😄 (and your work @ameknas-phac for starting this branch).

@sgsutcliffe the following dependencies are missing from `setup.py`: `pytest`, `scipy`, `psutil`. When you have time could you add them so installation with pip + testing works?

File: .github/workflows/branch.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+name: GAS branch protection
```
Comment (apetkau) [inline .github/workflows/branch.yml:1]: Thanks for adding this branch protection workflow 😄

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+jobs:
+  test:
+    name: Run pytest
+    # Only run on push if this is the phac-nml dev branch (merged PRs)
+    if: "${{ github.event_name != 'push' || (github.event_name == 'push' && github.repository == 'phac-nml/genomic_address_service') }}"
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.10", "3.11"]
```
Comment (apetkau) [inline .github/workflows/ci.yml:24]: I wonder if you should do python versions `3.10` and `3.13` (e.g., the min and max python versions)?

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+__pycache__
+test_GAS_assign.py
```
Comment (apetkau) [inline .gitignore:2]: Can you remove these @sgsutcliffe from `.gitignore` (the `test_*.py` entries)? I'm wondering if they would mess up with tracking changes of our tests

File: CHANGELOG.md [hunk 4]
```diff
@@ -3,6 +3,17 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Development
+
+- Fixed pytests
```
Comment (apetkau) [inline CHANGELOG.md:8]: These entries should be under `[0.1.3]` section since that hasn't been released yet.

**Review by apetkau — CHANGES_REQUESTED** [review]
Looks great. Thanks so much @sgsutcliffe. A few in-line comments.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (sgsutcliffe) [reply .github/workflows/ci.yml]: Python 3.13 causes a problem with numba

Comment (sgsutcliffe) [reply .github/workflows/ci.yml]: Looks like numba cannot handle python 3.13 [Release information](https://numba.readthedocs.io/en/stable/user/installing.html)

**Comment by sgsutcliffe** [comment]
For now I am going to run pytest in 3.10, 3.11 and 3.12 for the tests until they work, then I will drop 3.11.

**Review by apetkau — APPROVED** [review]
This looks great. Thanks so much 😄

**Pull request merged**